### PR TITLE
Composite measures

### PR DESCRIPTION
**Pull request by hackrm**
Pull requests into Bonnie Bundler require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

Updated bonnie_bundler so composite measures can now be loaded.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1750
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Code coverage has not gone down and all code touched or added is covered.
         Master branch coverage: 64.36% 
         composite_measures branch coverage: 67.04%
- [x] Automated regression test(s) pass - N/A

**Reviewer 1:**

Name: @mayerm94
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass NA
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree NA

**Reviewer 2:**

Name: @hossenlopp 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass N/A
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree N/A

### CODE REVIEW

File: lib/models/cql_measure.rb [hunk 1]
```diff
@@ -166,6 +168,21 @@ def count_expression_logical_branches(expression)
     self.complexity
   end
 
+  # Component measure hqmf_set_id is formatted as follows:
+  #   <composite_hqmf_set_id>&<component_hqmf_set_id>
+  def is_component
+    if self.hqmf_set_id.count("&") > 0
```
Comment (hossenlopp) [inline lib/models/cql_measure.rb]: The body of this function could simply be: ```self.hqmf_set_id.count("&") > 0```

File: lib/models/cql_measure.rb [hunk 2]
```diff
@@ -166,6 +168,21 @@ def count_expression_logical_branches(expression)
     self.complexity
   end
 
+  # Component measure hqmf_set_id is formatted as follows:
+  #   <composite_hqmf_set_id>&<component_hqmf_set_id>
+  def is_component
```
Comment (hossenlopp) [inline lib/models/cql_measure.rb]: To follow ruby naming guidelines for methods that return boolean, this method name should end with a question mark.

File: lib/measures/loading/cql_loader.rb [hunk 3]
```diff
@@ -1,62 +1,106 @@
...
-      # Grabs the cql file contents, the elm_xml contents, elm_json contents and the hqmf file path
-      files = get_files_from_zip(zip_file, out_dir)
-
-      # Load hqmf into HQMF Parser
-      hqmf_model = Measures::Loader.parse_hqmf_model(files[:HQMF_XML_PATH])
-
-      # Get main measure from hqmf parser
-      main_cql_library = hqmf_model.cql_measure_library
+    # Verifies contents of the given measure are valid (works for regular, composite and component measures)
+    def self.valid_measure_contents?(f_path)
```
Comment (hossenlopp) [inline lib/measures/loading/cql_loader.rb]: `f_path` should be a bit more descriptive in its naming.

File: lib/measures/loading/cql_loader.rb [hunk 4]
```diff
@@ -84,22 +128,181 @@ def self.set_data_criteria_code_list_ids(json, cql_artifacts)
...
+        if File.directory?(file)
+          component_measure = create_measure(file, current_user, measure_details, vsac_options, vsac_ticket_granting_ticket)
+          component_measures << component_measure
+        end
       end
+      component_measures
+    end
+
+    # Creates and returns a measure 
+    def self.create_measure(measure_dir, user, measure_details, vsac_options, vsac_ticket_granting_ticket, component_elm_files=nil)
```
Comment (hossenlopp) [inline lib/measures/loading/cql_loader.rb]: Why are the component_elm_files passed in here and used on line 221? It looks like line 213 fetches all files from the directory.

File: lib/measures/loading/cql_loader.rb [hunk 5]
```diff
@@ -84,22 +128,181 @@ def self.set_data_criteria_code_list_ids(json, cql_artifacts)
...
+
+      # Create CQL Measure
+      measure_details["composite"] = composite_measure?(measure_dir)
+      measure = Measures::Loader.load_hqmf_cql_model_json(json, user, cql_artifacts[:all_value_set_oids], main_cql_library, cql_artifacts[:cql_definition_dependency_structure],
+                                                          cql_artifacts[:elms], cql_artifacts[:elm_annotations], files[:CQL], measure_details, cql_artifacts[:value_set_oid_version_objects])
       measure
     end
 
+    def self.get_files_from_directory(dir)
+      cql_paths = Dir.glob(File.join("#{dir}/**.cql")).select
```
Comment (hossenlopp) [inline lib/measures/loading/cql_loader.rb]: I think `.select` not needed here since it is intended to take a block to filter the array. In this case no block is given.

Additionally, this could be where a `.sort` could be added to make the ordering of libraries consistent.

**Review by hossenlopp — CHANGES_REQUESTED** [review]

File: lib/measures/loading/cql_loader.rb [hunk 6]
```diff
@@ -84,22 +128,179 @@ def self.set_data_criteria_code_list_ids(json, cql_artifacts)
...
+        end
+      end # End of temporary directory usage 
+
+      # Put measure (and component measures) into an array to return
+      measures = component_measures << measure
+      return measures
+    end
+
+    # Creates a composite's component measures 
+    def self.create_component_measures(component_measures, current_directory, current_user, measure_details, vsac_options, vsac_ticket_granting_ticket)
```
Comment (hossenlopp) [inline lib/measures/loading/cql_loader.rb]: This method could be better encapsulated by creating the list of component_measures in the method and returning the list, instead passing in an empty array to fill.

File: test/unit/composite_cql_loader_test.rb [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+      dump_db
+      user = User.new
+      user.save
+
+      measure_details = { 'episode_of_care'=> false }
+
+      # Catch a NoMethodError because inside of cql_loader.rb in find_definition_in_elm
+      # The missing component path does not exist
+      # This error is handled in measures_controller.rb create
+      assert_raise NoMethodError do
```
Comment (hossenlopp) [inline test/unit/composite_cql_loader_test.rb]: This should be raising a `MeasureLoadingException`.

File: lib/models/cql_measure.rb [hunk 8]
```diff
@@ -22,6 +22,8 @@ class CqlMeasure
   field :continuous_variable, type: Boolean
   field :calculate_sdes, type: Boolean
   field :episode_ids, type: Array
+  field :composite, type: Boolean, default: false
+  field :components, type: Array, default: []
```
Comment (hossenlopp) [inline lib/models/cql_measure.rb]: This field would be more descriptive if named `:component_hqmf_set_ids`.

File: lib/models/cql_measure.rb [hunk 9]
```diff
@@ -166,6 +168,18 @@ def count_expression_logical_branches(expression)
     self.complexity
   end
 
+  # Component measure hqmf_set_id is formatted as follows:
+  #   <composite_hqmf_set_id>&<component_hqmf_set_id>
+  def is_component?
```
Comment (hossenlopp) [inline lib/models/cql_measure.rb]: Should be replaced with a field to mark it as a component instead of trusting the existence of the `&`.

File: lib/measures/loading/cql_loader.rb [hunk 10]
```diff
@@ -84,28 +133,181 @@ def self.set_data_criteria_code_list_ids(json, cql_artifacts)
...
+        # If it is a composite measure, load in each of the components
+        # Components must be loaded first so their elms can be passed onto the composite
+        if composite_measure?(current_directory)
+          component_measures = create_component_measures(current_directory, current_user, measure_details, vsac_options, vsac_ticket_granting_ticket)
+          component_measures.each do |component_measure|
+            component_elms[:ELM_JSON].push(*component_measure.elm) 
+            if (component_elms[:ELM_XML].keys & component_measure.elm_annotations.keys).count > 0 
+                puts "WARNING: Component measures share libraries. Version may or may not be the same."
+            end
+            component_elms[:ELM_XML] = component_measure.elm_annotations.merge(component_elms[:ELM_XML])
```
Comment (jbradl11) [inline lib/measures/loading/cql_loader.rb]: @mayerm94 Didn't we decide that `elm_annotations` for component measures would not be stored on the composite measure?

Comment (mayerm94) [reply lib/measures/loading/cql_loader.rb]: Yeah, i thought @zlister had a commit that removed this after the discussion about the comment here https://github.com/projecttacoma/bonnie_bundler/pull/167/commits/80f58c99680ce4d1a280fd5e0a0fec9b248613a2#r227127457 I do not see that commit though

Comment (mayerm94) [reply lib/measures/loading/cql_loader.rb]: I can't link to the comment for some reason. It says @zlister marked the comment as resolved

Comment (zlister) [reply lib/measures/loading/cql_loader.rb]: Hmm... I thought I did make that commit. Looking through the logs it doesn't look like I did. Fixing that now.

**Review by mayerm94 — APPROVED** [review]

File: Gemfile [hunk 11]
```diff
@@ -1,7 +1,7 @@
 source "https://rubygems.org"
 gemspec
 
-# gem 'health-data-standards', :git => 'https://github.com/projectcypress/health-data-standards.git', :branch => 'master_bonnie'
+gem 'health-data-standards', :git => 'https://github.com/projectcypress/health-data-standards.git', :branch => 'composite_measures'
```
Comment (hossenlopp) [inline Gemfile]: Can be updated to `bonnie-v3.1` branch now.

Comment (hossenlopp) [reply Gemfile]: The Gemfile.lock needs to be updated as well.

File: test/fixtures/vcr_cassettes/multi_library_webcalls.yml [hunk 12]
```diff
@@ -37,12 +37,12 @@ http_interactions:
       Content-Type:
```
Comment (hossenlopp) [inline test/fixtures/vcr_cassettes/multi_library_webcalls.yml:37]: This cassette should not have been re-recorded.

Comment (hossenlopp) [reply test/fixtures/vcr_cassettes/multi_library_webcalls.yml:37]: After discussion in the lab we decided to not worry about this.

**Review by hossenlopp — APPROVED** [review]
